ext/soap/soap.c

```diff
@@ -2769,27 +2785,28 @@ PHP_METHOD(SoapClient, __setCookie)
 	char *name;
 	char *val = NULL;
 	size_t  name_len, val_len = 0;
-	zval *cookies;
+	zval *cookies, rv;
 	zval *this_ptr = ZEND_THIS;
 
 	if (zend_parse_parameters(ZEND_NUM_ARGS(), "s|s!", &name, &name_len, &val, &val_len) == FAILURE) {
 		RETURN_THROWS();
 	}
 
 	if (val == NULL) {
-		if ((cookies = zend_hash_str_find(Z_OBJPROP_P(this_ptr), "_cookies", sizeof("_cookies")-1)) != NULL &&
-		    Z_TYPE_P(cookies) == IS_ARRAY) {
+		cookies = zend_read_property(soap_class_entry, Z_OBJ_P(this_ptr), "_cookies", sizeof("_cookies")-1, 1, &rv);
+		if (cookies != NULL && Z_TYPE_P(cookies) == IS_ARRAY) {
 			zend_hash_str_del(Z_ARRVAL_P(cookies), name, name_len);
 		}
 	} else {
 		zval zcookie;
 
-		if ((cookies = zend_hash_str_find(Z_OBJPROP_P(this_ptr), "_cookies", sizeof("_cookies")-1)) == NULL ||
-		    Z_TYPE_P(cookies) != IS_ARRAY) {
+		cookies = zend_read_property(soap_class_entry, Z_OBJ_P(this_ptr), "_cookies", sizeof("_cookies")-1, 1, &rv);
+		if (cookies == NULL || Z_TYPE_P(cookies) != IS_ARRAY) {
 			zval tmp_cookies;
 
 			array_init(&tmp_cookies);
-			cookies = zend_hash_str_update(Z_OBJPROP_P(this_ptr), "_cookies", sizeof("_cookies")-1, &tmp_cookies);
+			zend_update_property(soap_class_entry, Z_OBJ_P(this_ptr), "_cookies", sizeof("_cookies")-1, &tmp_cookies);
+			ZVAL_COPY(cookies, &tmp_cookies);
```

Comment:
I have problems here with the refcount of `cookies`. No matter what I try, it ends up being `2` (e.g. `ZVAL_EMPTY_ARRAY()` also initializes `refcount` to 2). Because of this `add_assoc_zval_ex()` fails with an assertion error (in `HT_ASSERT_RC1`) when running `ext/soap/tests/bug49898.phpt`.

What should I use then to update an array property with a new key?

Reply:
I think you'll want to do something like `Z_DELREF(tmp_cookie); cookies = Z_OBJ_P(this_ptr)->write_property(...)`.

Reply:
Thanks! It worked :) (but I haven't pushed it)